### DISCUSSION TITLE
CMSIS-DAP v2 USB protocol

### DISCUSSION THREAD

**Discussion by diggit** [ARMmbed/DAPLink#955] [Q&A]
Hi,
I am trying to understand how does CMSIS-DAP USB protocol work.
I am not aware of way, how to determine length of command payload without understanding command itself.
v1 (HID) transport overcomes this issue by transferring 1 command/response per HID frame.

1. Does v2 use same concept, but instead of HID frames, USB bulk frames are used?

**Comment by rgrr** [comment]
Concerning your length question: I had the same question on my own.  I ended up writing a function which calculates the length of a request.  Thanks to the protocol which does not include packet length information.

If you are still interested, I can post the module.

**Comment by diggit** [comment]
I don't think, my question is clear or maybe I don't understand your reply.

I am trying to find info about how the receiving side recognize boundaries of packets in incoming data stream. Eg. host wants to send 3 packets. Does each of them have to get into separate bulk xfer (easy for receiver, but wastes bandwidths) or are they somehow packed together in one larger bulk xfer? If the second case, how receiving side unpacks them?

**Reply by rgrr** [reply]
Ok, then I've misunderstood your question.  My former problem was that I wanted to check the incoming data stream for complete commands.  It is for me unclear how commands are received (in one big chunk or in parts) in DAPv2, especially if DAP_PACKET_SIZE is set to something other then 64.
On the tool request side I'ld assume, that it's tool dependent if it packs two or more requests into a single protocol packet.
And at least (if there is any tool which actually does that) I'ld expect that the tool should be capable to receive multiple responses in one packet.
But these are assumptions ;-)
